SSA/Projects/FullyHomomorphicEncryption/Basic.lean

```diff
@@ -43,6 +48,62 @@ theorem WithBot.npow_coe_eq_npow (n : Nat) (x : ℕ) : (WithBot.some x : WithBot
 
 noncomputable def f : (ZMod q)[X] := X^(2^n) + 1
 
+-- theorem zmodq_eq_finq : ZMod q = Fin q := by
+--   have h : q > 1 := hqgt1.elim
+--   unfold ZMod
+--   cases q
+--   · exfalso
+--     apply Nat.not_lt_zero 1
+--     exact h
+--   · simp [Fin]
+--   done
+
+def ZMod.toInt (x : ZMod q) : Int := ZMod.cast x
+def ZMod.toFin (x : ZMod q) : Fin q := ZMod.cast x
+
+theorem ZMod.toInt_coe_eq (x : ZMod q) : ↑(x.toInt) = x := by
+  unfold toInt
+  simp
+
+theorem ZMod.cast_eq_val' [Fact (q > 1)](x : ZMod q) : @cast ℤ Ring.toAddGroupWithOne q x = (val x : ℕ) := by
```

Comment:
I don't understand the statement. Also you can use type ascriptions to make sure that you trigger the right coercion and maybe use two type ascriptions for two coercions e.g. `((x : Nat) : Int)`